sqlhooks.go

```diff
@@ -149,6 +160,99 @@ func (conn *ExecerContext) Exec(query string, args []driver.Value) (driver.Resul
 	return nil, errors.New("Exec was called when ExecContext was implemented")
 }
 
+// QueryerContext implements a database/sql.driver.QueryerContext
+type QueryerContext struct {
+	*Conn
+}
+
+func isQueryer(conn driver.Conn) bool {
+	switch conn.(type) {
+	case driver.QueryerContext:
+		return true
+	case driver.Queryer:
+		return true
+	default:
+		return false
+	}
+}
+
+func (conn *QueryerContext) queryContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Rows, error) {
+	switch c := conn.Conn.Conn.(type) {
+	case driver.QueryerContext:
+		return c.QueryContext(ctx, query, args)
+	case driver.Queryer:
+		dargs, err := namedValueToValue(args)
+		if err != nil {
+			return nil, err
+		}
+		return c.Query(query, dargs)
+	default:
+		// This should not happen
+		return nil, errors.New("QueryerContext created for a non Queryer driver.Conn")
+	}
+}
+
+func (conn *QueryerContext) QueryContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Rows, error) {
+	var err error
+
+	list := namedToInterface(args)
+
+	// Query `Before` Hooks
+	if ctx, err = conn.hooks.Before(ctx, query, list...); err != nil {
+		return nil, err
+	}
+
+	results, err := conn.queryContext(ctx, query, args)
+	if err != nil {
+		return results, handlerErr(ctx, conn.hooks, err, query, list...)
+	}
+
+	if ctx, err = conn.hooks.After(ctx, query, list...); err != nil {
+		return nil, err
+	}
+
+	return results, err
+}
+
+// ExecerQueryerContext implements database/sql.driver.ExecerContext and
+// database/sql.driver.QueryerContext
+type ExecerQueryerContext struct {
+	*Conn
+	*ExecerContext
+	*QueryerContext
+}
+
+// ExecerQueryerContext implements database/sql.driver.ExecerContext and
+// database/sql.driver.QueryerContext
+type ExecerQueryerContextWithSessionResetter struct {
+	*Conn
+	*ExecerContext
+	*QueryerContext
+	*SessionResetter
+}
+
+type SessionResetter struct {
+	*Conn
+}
+
+func isSessionResetter(conn driver.Conn) bool {
+	switch conn.(type) {
+	case driver.SessionResetter:
+		return true
+	default:
+		return false
+	}
+}
+
+func (s *SessionResetter) ResetSession(ctx context.Context) error {
+	switch c := s.Conn.Conn.(type) {
```

Comment:
do we need this optional type assert. Should we just assert it to driver.SessionResetter since we should only be using this if Conn is a driver.SessionResetter. I worry about accidently using this and causing weird bugs were a client calls ResetSession but we end up no-opping.

Reply:
Good point, I have modified it to assert